dpctl/tensor/_manipulation_functions.py

```diff
@@ -741,6 +741,107 @@ def finfo(dtype):
+
+    Returns:
+        out (usm_narray): A tuple of arrays.
+
+    Raises:
+        AxisError: if provided axis position is invalid.
+    """
+    if not isinstance(X, dpt.usm_ndarray):
+        raise TypeError(f"Expected usm_ndarray type, got {type(X)}.")
+
+    axis = normalize_axis_index(axis, X.ndim)
+    Y = dpt.moveaxis(X, axis, 0)
+
+    return tuple(Y[i] for i in range(Y.shape[0]))
+
+
+def moveaxis(X, src, dst):
+    """
+    Args:
+        x (usm_ndarray): input array
+
+        src (int or a sequence of int): Original positions of the axes to move.
+        These must be unique. If `X` has rank (i.e., number of dimensions) `N`,
+        a valid `axis` must reside in the half-open interval `[-N, N)`.
+
+        dst (int or a sequence of int): Destination positions for each of the
+        original axes. These must also be unique. If `X` has rank
+        (i.e., number of dimensions) `N`, a valid `axis` must reside
+        in the half-open interval `[-N, N)`.
+
+    Returns:
+        out (usm_narray): Array with moved axes.
+            The returned array must has the same data type as `X`,
+            is created on the same device as `X` and has the same USM allocation
+            type as `X`.
+
+    Raises:
+        AxisError: if provided axis position is invalid.
+    """
+    if not isinstance(X, dpt.usm_ndarray):
+        raise TypeError(f"Expected usm_ndarray type, got {type(X)}.")
+
+    if not isinstance(src, (tuple, list)):
+        src = (src,)
+
+    if not isinstance(dst, (tuple, list)):
+        dst = (dst,)
+
+    src = normalize_axis_tuple(src, X.ndim, "src")
+    dst = normalize_axis_tuple(dst, X.ndim, "dst")
+    ind = list(range(0, X.ndim))
+    for i in range(len(src)):
+        ind.remove(src[i])  # using the value here which is the same as index
+        ind.insert(dst[i], src[i])
+
+    return dpt.permute_dims(X, tuple(ind))
+
+
+def swapaxes(X, axis1, axis2):
+    """
+    Args:
+        x (usm_ndarray): input array
+
+        axis1 (int): First axis.
+        If `X` has rank (i.e., number of dimensions) `N`,
+        a valid `axis` must reside in the half-open interval `[-N, N)`.
+
+        axis2 (int): Second axis.
+        If `X` has rank (i.e., number of dimensions) `N`,
+        a valid `axis` must reside in the half-open interval `[-N, N)`.
+
+    Returns:
+        out (usm_narray): Swapped array.
+            The returned array must has the same data type as `X`,
+            is created on the same device as `X` and has the same USM allocation
+            type as `X`.
+
+    Raises:
+        AxisError: if provided axis position is invalid.
+    """
+    if not isinstance(X, dpt.usm_ndarray):
+        raise TypeError(f"Expected usm_ndarray type, got {type(X)}.")
+
+    axis1 = normalize_axis_index(axis1, X.ndim, "axis1")
+    axis2 = normalize_axis_index(axis2, X.ndim, "axis2")
+
+    ind = list(range(0, X.ndim))
+    ind[axis1] = axis2
+    ind[axis2] = axis1
+    return dpt.permute_dims(X, tuple(ind))
```

Comment:
Instead of creating `list` to be able to mutate it only to then copy it into `tuple`, use generator to create the right tuple directly. Here is an example:

```
In [1]: def permuted_range(n, a1, a2):
   ...:     for i in range(n):
   ...:         if i == a1:
   ...:             yield a2
   ...:         elif i == a2:
   ...:             yield a1
   ...:         else:
   ...:             yield i
   ...:

In [2]: tuple(permuted_range(5, 1, 3))
Out[2]: (0, 3, 2, 1, 4)
```

Reply:
Actually, the way it is currently implement is faster:

```

In [13]: def way1(n, a1, a2):
    ...:     ind = list(range(n))
    ...:     ind[a1] = a2
    ...:     ind[a2] = a1
    ...:     return tuple(ind)
    ...:

In [14]: def permuted_range(n, a1, a2):
    ...:     for i in range(n):
    ...:         if i == a1:
    ...:             yield a2
    ...:         elif i == a2:
    ...:             yield a1
    ...:         else:
    ...:             yield i
    ...:

In [15]: def way2(n, a1, a2):
    ...:     return tuple(permuted_range(n, a1, a2))
    ...:

In [16]: way1(5, 1, 3) == way2(5, 1, 3)
Out[16]: True

In [17]: %timeit way1(5, 1, 3)
434 ns ± 23 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [18]: %timeit way2(5, 1, 3)
865 ns ± 98 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```